### ISSUE TITLE
__DATA__ conflicts with Proc::Daemon. A "half closed" filehandle.

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#10130]
Migrated from [rt.perl.org#72526](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=72526) (status was 'rejected')

Searchable as RT72526$

**Comment by p5pRT** [comment]
### From kissg@ssg.ki.iif.hu


This is a bug report for perl from kissg@&#8203;ssg\.ki\.iif\.hu\,
generated with the help of perlbug 1\.36 running under perl 5\.10\.0\.




---
Please run this shell script that calls the PERL program below&#8203;:

\-\-\-\-\-\-\-\-\-\-\-\-\-\-8\<\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-8\<\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
\#/bin/sh

echo "No fork"
\./perlbug
cat \./perlbug\.out

echo
echo "Daemonize process before opening files"
\./perlbug \-fork
sleep 2                 \# allow child process to finish
cat \./perlbug\.out
\-\-\-\-\-\-\-\-\-\-\-\-\-\-8\<\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-8\<\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-

File perlbug&#8203;:
\-\-\-\-\-\-\-\-\-\-\-\-\-\-8\<\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-8\<\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
\#\!/usr/bin/perl \-s

use Proc&#8203;::Daemon;
use IO&#8203;::File;
our $fork;

system\("\(echo before open pid=$$&#8203;: ; ls \-l /proc/$$/fd\) > /tmp/perlbug\.out"\);
Proc&#8203;::Daemon&#8203;::Init if $fork;

\{
&nbsp;        my $fh1 = IO&#8203;::File\->new\('/etc/passwd'\); \# file is open
&nbsp;        my $fh2 = IO&#8203;::File\->new\('/etc/fstab'\);  \# file is open
&nbsp;        my $fh3 = IO&#8203;::File\->new\('/etc/inittab'\);\# file is open
&nbsp;        system\("\(echo after open pid=$$&#8203;:;ls \-l /proc/$$/fd\)>>/tmp/perlbug\.out"\);
\}
\# $fh1\, $fh2\, $fh3 objects are destroyed\. Files must be closed\.
\# Let's check it&#8203;:
system\("\(echo after destroy pid=$$&#8203;: ; ls \-l /proc/$$/fd\) >> /tmp/perlbug\.out"\);

\# comment out this line&#8203;:
\_\_DATA\_\_
\-\-\-\-\-\-\-\-\-\-\-\-\-\-8\<\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-8\<\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-

The result is this&#8203;:
\-\-\-\-\-\-\-\-\-\-\-\-\-\-8\<\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-8\<\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
kissg@&#8203;bakacsin&#8203;:/tmp$ \./perlbug\.sh
No fork
before open pid=31200&#8203;:
total 0
lrwx\-\-\-\-\-\- 1 kissg kissg 64 2010\-02\-04 09&#8203;:25 0 \-> /dev/pts/9
lrwx\-\-\-\-\-\- 1 kissg kissg 64 2010\-02\-04 09&#8203;:25 1 \-> /dev/pts/9
lrwx\-\-\-\-\-\- 1 kissg kissg 64 2010\-02\-04 09&#8203;:25 2 \-> /dev/pts/9
lr\-x\-\-\-\-\-\- 1 kissg kissg 64 2010\-02\-04 09&#8203;:25 3 \-> /tmp/perlbug
lr\-x\-\-\-\-\-\- 1 kissg kissg 64 2010\-02\-04 09&#8203;:25 4 \-> pipe&#8203;:\[6127332\]
after open pid=31200&#8203;:
total 0
lrwx\-\-\-\-\-\- 1 kissg kissg 64 2010\-02\-04 09&#8203;:25 0 \-> /dev/pts/9
lrwx\-\-\-\-\-\- 1 kissg kissg 64 2010\-02\-04 09&#8203;:25 1 \-> /dev/pts/9
lrwx\-\-\-\-\-\- 1 kissg kissg 64 2010\-02\-04 09&#8203;:25 2 \-> /dev/pts/9
lr\-x\-\-\-\-\-\- 1 kissg kissg 64 2010\-02\-04 09&#8203;:25 3 \-> /tmp/perlbug
lr\-x\-\-\-\-\-\- 1 kissg kissg 64 2010\-02\-04 09&#8203;:25 4 \-> /etc/passwd
lr\-x\-\-\-\-\-\- 1 kissg kissg 64 2010\-02\-04 09&#8203;:25 5 \-> /etc/fstab
lr\-x\-\-\-\-\-\- 1 kissg kissg 64 2010\-02\-04 09&#8203;:25 6 \-> /etc/inittab
lr\-x\-\-\-\-\-\- 1 kissg kissg 64 2010\-02\-04 09&#8203;:25 7 \-> pipe&#8203;:\[6127345\]
after destroy pid=31200&#8203;:
total 0
lrwx\-\-\-\-\-\- 1 kissg kissg 64 2010\-02\-04 09&#8203;:25 0 \-> /dev/pts/9
lrwx\-\-\-\-\-\- 1 kissg kissg 64 2010\-02\-04 09&#8203;:25 1 \-> /dev/pts/9
lrwx\-\-\-\-\-\- 1 kissg kissg 64 2010\-02\-04 09&#8203;:25 2 \-> /dev/pts/9
lr\-x\-\-\-\-\-\- 1 kissg kissg 64 2010\-02\-04 09&#8203;:25 3 \-> /tmp/perlbug
lr\-x\-\-\-\-\-\- 1 kissg kissg 64 2010\-02\-04 09&#8203;:25 4 \-> pipe&#8203;:\[6127355\]

Daemonize process before opening files
before open pid=31211&#8203;:
total 0
lrwx\-\-\-\-\-\- 1 kissg kissg 64 2010\-02\-04 09&#8203;:25 0 \-> /dev/pts/9
lrwx\-\-\-\-\-\- 1 kissg kissg 64 2010\-02\-04 09&#8203;:25 1 \-> /dev/pts/9
lrwx\-\-\-\-\-\- 1 kissg kissg 64 2010\-02\-04 09&#8203;:25 2 \-> /dev/pts/9
lr\-x\-\-\-\-\-\- 1 kissg kissg 64 2010\-02\-04 09&#8203;:25 3 \-> /tmp/perlbug
lr\-x\-\-\-\-\-\- 1 kissg kissg 64 2010\-02\-04 09&#8203;:25 4 \-> pipe&#8203;:\[6127368\]
after open pid=31216&#8203;:
total 0
lrwx\-\-\-\-\-\- 1 kissg kissg 64 2010\-02\-04 09&#8203;:25 0 \-> /dev/null
lrwx\-\-\-\-\-\- 1 kissg kissg 64 2010\-02\-04 09&#8203;:25 1 \-> /dev/null
lrwx\-\-\-\-\-\- 1 kissg kissg 64 2010\-02\-04 09&#8203;:25 2 \-> /dev/null
lr\-x\-\-\-\-\-\- 1 kissg kissg 64 2010\-02\-04 09&#8203;:25 3 \-> /etc/passwd
lr\-x\-\-\-\-\-\- 1 kissg kissg 64 2010\-02\-04 09&#8203;:25 4 \-> /etc/fstab
lr\-x\-\-\-\-\-\- 1 kissg kissg 64 2010\-02\-04 09&#8203;:25 5 \-> /etc/inittab
lr\-x\-\-\-\-\-\- 1 kissg kissg 64 2010\-02\-04 09&#8203;:25 6 \-> pipe&#8203;:\[6127381\]
after destroy pid=31216&#8203;:
total 0
lrwx\-\-\-\-\-\- 1 kissg kissg 64 2010\-02\-04 09&#8203;:25 0 \-> /dev/null
lrwx\-\-\-\-\-\- 1 kissg kissg 64 2010\-02\-04 09&#8203;:25 1 \-> /dev/null
lrwx\-\-\-\-\-\- 1 kissg kissg 64 2010\-02\-04 09&#8203;:25 2 \-> /dev/null
lr\-x\-\-\-\-\-\- 1 kissg kissg 64 2010\-02\-04 09&#8203;:25 3 \-> /etc/passwd
lr\-x\-\-\-\-\-\- 1 kissg kissg 64 2010\-02\-04 09&#8203;:25 4 \-> pipe&#8203;:\[6127397\]
/tmp$ 
\-\-\-\-\-\-\-\-\-\-\-\-\-\-8\<\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-8\<\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-

Note that /etc/passwd \(filehandle 3\) remained open at the second run
even after the associated IO&#8203;::Handle object $fh1 was destroyed\.

However if you comment out the \_\_DATA\_\_ line everything works well\.

The problem is not limited to ordinary files\.
Network sockets are also affected\.
Actually this is the way I found the problem\.

A server process that daemonize itself with Proc&#8203;::Daemon&#8203;::Init\(\)
does not close the TCP connection coming from the \_first\_ client\.
The daemon uses Device&#8203;::USB module that contains \_\_DATA\_\_ and
inline C code\. At compile time the interpreter reopens USB\.pm as 5
and never closes it\. Then Proc&#8203;::Daemon&#8203;::Init\(\) forks and closes
all handles including 5\. The first accept\(\) returns 5 as the first
free number but when all event handlers finish their jobs and all of
them unreferences the IO&#8203;::Handle object\, "somebody" keeps it open\.
Subsequent clients are connected via 6\,7\,8 etc and they
are closed well when object's reference reaches zero\.

\(Yes\, I could find a workaround to handle this situation\.\)
As far as I can see the real problem is that upgrade of
a "third party" module can break well working old programs
if the author of the module starts to use \_\_DATA\_\_ for any reason\.

My Linux system is a Debian lenny with kernel 2\.6\.26\-2\-686\.
The bug is also reproduced on a 5 year old Red Hat machine with
2\.4\.21\-32\.ELsmp kernel and PERL 5\.8\.0\.

Gabor



---


---
Flags&#8203;:
&nbsp;    category=core
&nbsp;    severity=low


---
Site configuration information for perl 5\.10\.0&#8203;:

Configured by Debian Project at Fri Aug 28 22&#8203;:30&#8203;:10 UTC 2009\.

Summary of my perl5 \(revision 5 version 10 subversion 0\) configuration&#8203;:
&nbsp;  Platform&#8203;:
&nbsp;    osname=linux\, osvers=2\.6\.26\-2\-amd64\, archname=i486\-linux\-gnu\-thread\-multi
&nbsp;    uname='linux puccini 2\.6\.26\-2\-amd64 \#1 smp fri aug 14 07&#8203;:12&#8203;:04 utc 2009 i686 gnulinux '
&nbsp;    config\_args='\-Dusethreads \-Duselargefiles \-Dccflags=\-DDEBIAN \-Dcccdlflags=\-fPIC \-Darchname=i486\-linux\-gnu \-Dprefix=/usr \-Dprivlib=/usr/share/perl/5\.10 \-Darchlib=/usr/lib/perl/5\.10 \-Dvendorprefix=/usr \-Dvendorlib=/usr/share/perl5 \-Dvendorarch=/usr/lib/perl5 \-Dsiteprefix=/usr/local \-Dsitelib=/usr/local/share/perl/5\.10\.0 \-Dsitearch=/usr/local/lib/perl/5\.10\.0 \-Dman1dir=/usr/share/man/man1 \-Dman3dir=/usr/share/man/man3 \-Dsiteman1dir=/usr/local/man/man1 \-Dsiteman3dir=/usr/local/man/man3 \-Dman1ext=1 \-Dman3ext=3perl \-Dpager=/usr/bin/sensible\-pager \-Uafs \-Ud\_csh \-Ud\_ualarm \-Uusesfio \-Uusenm \-DDEBUGGING=\-g \-Doptimize=\-O2 \-Duseshrplib \-Dlibperl=libperl\.so\.5\.10\.0 \-Dd\_dosuid \-des'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    useithreads=define\, usemultiplicity=define
&nbsp;    useperlio=define\, d\_sfio=undef\, uselargefiles=define\, usesocks=undef
&nbsp;    use64bitint=undef\, use64bitall=undef\, uselongdouble=undef
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'\, ccflags ='\-D\_REENTRANT \-D\_GNU\_SOURCE \-DDEBIAN \-fno\-strict\-aliasing \-pipe \-I/usr/local/include \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'\,
&nbsp;    optimize='\-O2 \-g'\,
&nbsp;    cppflags='\-D\_REENTRANT \-D\_GNU\_SOURCE \-DDEBIAN \-fno\-strict\-aliasing \-pipe \-I/usr/local/include'
&nbsp;    ccversion=''\, gccversion='4\.3\.2'\, gccosandvers=''
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=1234
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=12
&nbsp;    ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8
&nbsp;    alignbytes=4\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='cc'\, ldflags =' \-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /lib /usr/lib /usr/lib64
&nbsp;    libs=\-lgdbm \-lgdbm\_compat \-ldb \-ldl \-lm \-lpthread \-lc \-lcrypt
&nbsp;    perllibs=\-ldl \-lm \-lpthread \-lc \-lcrypt
&nbsp;    libc=/lib/libc\-2\.7\.so\, so=so\, useshrplib=true\, libperl=libperl\.so\.5\.10\.0
&nbsp;    gnulibc\_version='2\.7'
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-Wl\,\-E'
&nbsp;    cccdlflags='\-fPIC'\, lddlflags='\-shared \-O2 \-g \-L/usr/local/lib'

Locally applied patches&#8203;:
&nbsp;    



---
@&#8203;INC for perl 5\.10\.0&#8203;:
&nbsp;    /etc/perl
&nbsp;    /usr/local/lib/perl/5\.10\.0
&nbsp;    /usr/local/share/perl/5\.10\.0
&nbsp;    /usr/lib/perl5
&nbsp;    /usr/share/perl5
&nbsp;    /usr/lib/perl/5\.10
&nbsp;    /usr/share/perl/5\.10
&nbsp;    /usr/local/lib/site\_perl
&nbsp;    \.



---
Environment for perl 5\.10\.0&#8203;:
&nbsp;    HOME=/home/kissg
&nbsp;    LANG=en\_US
&nbsp;    LANGUAGE \(unset\)
&nbsp;    LD\_LIBRARY\_PATH \(unset\)
&nbsp;    LOGDIR \(unset\)
&nbsp;    PATH=/home/kissg/bin&#8203;:/usr/local/bin&#8203;:/usr/bin&#8203;:/bin&#8203;:/usr/bin/X11&#8203;:/usr/games
&nbsp;    PERL\_BADLANG \(unset\)
&nbsp;    SHELL=/bin/bash

**Comment by p5pRT** [comment]
### From @iabyn

On Thu\, Feb 04\, 2010 at 01&#8203;:29&#8203;:18AM \-0800\, kissg@&#8203;ssg\.ki\.iif\.hu wrote&#8203;:
> Note that /etc/passwd \(filehandle 3\) remained open at the second run
> even after the associated IO&#8203;::Handle object $fh1 was destroyed\.

This isn't a problem with perl\. Proc&#8203;::Daemon is "secretly" closing file
descriptor 3 \(the one perl has associated with DATA\) without telling perl\.
When perl is asked to open a new file\, the OS returns fd 3\, so there are
now two perl handles associated with fd 3&#8203;: DATA and $fh1\. You ask perl to
close one of the handles\, but there's now two perl file handles associated
with fd 3\, so it just decrements the refcount rather than closing fh3\.
This leaves DATA associated with fh 3\, which is still opened as
/etc/passwd\.

You can see a similar effect withe these two programs that don't require
Proc&#8203;::Daemon&#8203;:

&nbsp;    use POSIX \(\);
&nbsp;    POSIX&#8203;::close\(3\);
&nbsp;    open my $fh\, '\<'\, '/etc/passwd' or die "open&#8203;: $\!\\n";
&nbsp;    close $fh;
&nbsp;    system "ls \-l /proc/$$/fd";
&nbsp;    \_\_DATA\_\_

and

&nbsp;    use POSIX \(\);
&nbsp;    open my $data\, $0\, or die;
&nbsp;    POSIX&#8203;::close\(3\);
&nbsp;    open my $fh\, '\<'\, '/etc/passwd' or die "open&#8203;: $\!\\n";
&nbsp;    close $fh;
&nbsp;    system "ls \-l /proc/$$/fd";



\-\- 
The warp engines start playing up a bit\, but seem to sort themselves out
after a while without any intervention from boy genius Wesley Crusher\.
&nbsp;    \-\- Things That Never Happen in "Star Trek" \#17

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From kissg@ssg.ki.iif.hu

I think if you are right\, there is no way to daemonize a process
in a correct way without exec\(\)\-ing another Perl interpreter and
replaying all the initialization phase\.
\(Corollary&#8203;: Proc&#8203;::Daemon has to be forgotten\.\)

An application any time can use a third party module with a
hidden \_\_DATA\_\_ segment\. So programmer has no way to figure
out which are the decriptors that must not be closed\.
In general case \(s\)he cannot get a list of open files\.

Did I miss something?

Gabor

**Comment by p5pRT** [comment]
### From deti@cpan.org

On Fr\. 05\. Feb\. 2010\, 15&#8203;:52&#8203;:56\, davem wrote&#8203;:
> This isn't a problem with perl\. Proc&#8203;::Daemon is "secretly" closing 
file
> descriptor 3 \(the one perl has associated with DATA\) without telling 
perl\.

1\) Not a perl problem? Isn't POSIX a part of perl nowadays?
2\) Why isn't it documented \(http&#8203;://perldoc\.perl\.org/POSIX\.html\) that 
POSIX&#8203;::close\( $FD \) does not close the real file descriptor?
3\) Does POSIX&#8203;::close\( $FD \) have any function at all if it doesn't tell 
perl that the file must be closed?
4\) How can I close a file knowing only the file descriptor\, not knowing 
the file handle? Or\, how can I get the file handle to \<CORE&#8203;::close
\( $FH \)> if I only know the file descriptor number\. For example if I 
want to close all file descriptors from 0 \.\. POSIX&#8203;::sysconf
\( POSIX&#8203;::\_SC\_OPEN\_MAX \) as Proc&#8203;::Daemon does to detach from the parent 
process\.

Thanks and regards

Detlef Pilzecker

**Comment by p5pRT** [comment]
### From @iabyn

On Sat\, May 21\, 2011 at 11&#8203;:32&#8203;:29AM \-0700\, Detlef Pilzecker via RT wrote&#8203;:
> On Fr\. 05\. Feb\. 2010\, 15&#8203;:52&#8203;:56\, davem wrote&#8203;:
> > This isn't a problem with perl\. Proc&#8203;::Daemon is "secretly" closing 
> file
> > descriptor 3 \(the one perl has associated with DATA\) without telling 
> perl\.
> 
> 1\) Not a perl problem? Isn't POSIX a part of perl nowadays?
> 2\) Why isn't it documented \(http&#8203;://perldoc\.perl\.org/POSIX\.html\) that 
> POSIX&#8203;::close\( $FD \) does not close the real file descriptor?

It's documented to do what it does&#8203;: close the OS\-level file descriptor;
i\.e\. it's a thin wrapper over the close\(2\) system call\.

> 3\) Does POSIX&#8203;::close\( $FD \) have any function at all if it doesn't tell 
> perl that the file must be closed?

It's useful if you just want to close a UNIX file descriptor\.

> 4\) How can I close a file knowing only the file descriptor\, not knowing 
> the file handle? Or\, how can I get the file handle to \<CORE&#8203;::close
> \( $FH \)> if I only know the file descriptor number\. For example if I 
> want to close all file descriptors from 0 \.\. POSIX&#8203;::sysconf
> \( POSIX&#8203;::\_SC\_OPEN\_MAX \) as Proc&#8203;::Daemon does to detach from the parent 
> process\.

There isn't an easy way\. Bear in mind that several different Perl\-level
file handles\, spread over multiple threads\, may all map to the same OS
file descriptor\.

Some approaches to making this available at the perl level would be&#8203;:

1\) at the bottom&#8203;: add an extra level of indirection between the PerlIO
stack and the file descriptor table; this would allow file descriptors to
be "invalidated" rather than being accidentally reallocated\. Forcible
closing at this level would mean that all buffer flushing\, tied CLOSE\(\)
calling etc would be skipped\, similar to what's happening now with
POSIX&#8203;::close\(\)\.

2\) At the middle&#8203;: call the close method on every file stack in PL\_perlio;
this ensures that all buffers are flushed etc\, but has the insurmountable
drawback that only handles in the current thread are closed\.

3\) At the top&#8203;: scan the SV arenas for PVIOs\, and close each one\. This has
the same advantages and disadvantages as \(2\)\, except that in addition it
will call CLOSE\(\) on tied handles\, but there's no guarantee that all
PerlIO objects are linked form PVIOs\, so some handles might be missed\.

Note that the standard C library suffers from exactly the same problem&#8203;:

&nbsp;    main \(int argc\, char \*\*argv\, char \*\*env\)
&nbsp;    \{
&nbsp;    FILE \*f\, \*g;
&nbsp;    char s\[1000\];;
&nbsp;    f = fopen\("/etc/passwd"\, "r"\);
&nbsp;    close\(3\);
&nbsp;    g = fopen\("/etc/rpc"\, "r"\);
&nbsp;    fgets\(s\, 100\, f\);
&nbsp;    printf\("s=\[%s\]\\n"\, s\);  /\* prints a line from rpc\, not passwd \*/
&nbsp;    \}

In short\, there isn't a practical way to do what you want \(close all files
without side\-effects\)\.

\-\- 
"There's something wrong with our bloody ships today\, Chatfield\."
&nbsp;    \-\- Admiral Beatty at the Battle of Jutland\, 31st May 1916\.

**Comment by p5pRT** [comment]
@iabyn - Status changed from 'open' to 'rejected'
